smart_open/tests/test_smart_open.py

```diff
@@ -200,93 +202,158 @@ def test_http_pass(self):
+        self._test_compressed_http(".gz", True)
 
-        with open(test_file, 'rb') as infile:
-            compressed_data = infile.read()
+    def test_http_bz2_query(self):
+        """Can open bzip2 via http with a query appended to URI?"""
+        self._test_compressed_http(".bz2", True)
 
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
+    def test_http_xz_query(self):
+        """Can open xz via http with a query appended to URI?"""
+        self._test_compressed_http(".xz", True)
 
-        responses.add(responses.GET, "http://127.0.0.1/data.xz", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.xz")
 
-        # decompress the xz and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
+class SmartOpenFileObjTest(unittest.TestCase):
+    """
+    Test passing raw file objects.
+    """
+    text = 'Hello, world!'
 
+    def patch_string_io(func):
+        if not PY2:
+            # Python 3 has already got us covered
+            return func
+
+        def enter(self):
+            return self
+
+        def exit(self, exc_type, exc_val, exc_tb):
+            self.close()
+
+        @functools.wraps(func)
+        def wrapped_patch_string_io(self):
+            six.StringIO.__enter__ = enter
+            six.StringIO.__exit__ = exit
+            try:
+                return func(self)
+            finally:
+                del six.StringIO.__enter__
+                del six.StringIO.__exit__
+
+        return wrapped_patch_string_io
+
+    @patch_string_io
+    def test_read_bytes(self):
+        """Can we read bytes from a byte stream?"""
+        buffer = six.BytesIO(self.text.encode())
+        with smart_open.smart_open(buffer, 'rb') as sf:
+            data = sf.read()
+        self.assertEqual(data, self.text.encode())
+
+    @patch_string_io
+    def test_write_bytes(self):
+        """Can we write bytes to a byte stream?"""
+        buffer = six.BytesIO()
+        with smart_open.smart_open(buffer, 'wb') as sf:
+            sf.write(self.text.encode())
+            self.assertEqual(buffer.getvalue(), self.text.encode())
+
+    @unittest.skipIf(PY2, "Python 2 does not make a difference between str and bytes")
+    @patch_string_io
+    def test_read_str(self):
+        """Can we read strings from a text stream?"""
+        buffer = six.StringIO(self.text)
+        with smart_open.smart_open(buffer, 'r') as sf:
+            self.assertRaises(TypeError, sf.read)  # we expect binary mode
+
+    @unittest.skipIf(PY2, "Python 2 does not make a difference between str and bytes")
+    @patch_string_io
+    def test_write_str(self):
+        """Can we write strings to a text stream?"""
+        buffer = six.StringIO()
+        with smart_open.smart_open(buffer, 'w') as sf:
+            self.assertRaises(TypeError, sf.write, self.text)  # we expect binary mode
```

Comment:
What's the actual problem here?

The buffer is text. You've opened it in write text mode. You're writing text to it. Why should the test fail?

The name of the test and the docstring may be wrong, too.

Reply:
The problem is that `smart_open` always assumes that the wrapped stream is a binary stream. We do not support writing to text streams, and this test checks that.

Reply:
I see. Let's use a more informative test name and docstring here.

```python
def test_write_text_stream_fails(self):
    """smart_open does not support writing to a text stream.""" 
```

Reply:
> The problem is that `smart_open` always assumes that the wrapped stream is a binary stream.

Is this by necessity? Or why do we assume that?

Reply:
If it’s a text stream, there’s no new functionality that smart_open can add to it. Reading and writing is possible, but it’d be the same as doing stream.read() and stream.write() directly.

Reply:
Besides, there is no way to check if the wrapped stream is binary or text.